File: ifif/python/S1_16926.py
import sys


sys.stdin = open("input.txt", "r")

N, M, R = map(int, input().split())
arr = list( list(map(int, sys.stdin.readline().split())) for _ in range(N))
direction = [(1,0), (0,1), (-1,0), (0,-1)]

for _ in range(R):
    for r in range(min(N,M)//2):
        idx = 0
        x, y = r, r
        curr = arr[r][r]
        for d in direction:
            while True:
                dx = x+d[0]
                dy = y+d[1]
                if(r<=dx<N-r) and (r<=dy<M-r):
                    x, y = dx, dy
                    arr[x][y], curr = curr, arr[x][y]
                else:
                    break
for ar in arr:
    print(*ar)

File: ifif/java/S1_16926.java
import java.io.*;
import java.util.StringTokenizer;

public class S1_16926 {
    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("input.txt"));
        BufferedReader sc = new BufferedReader(new InputStreamReader(System.in));
        String[] NMR = sc.readLine().split(" ");
        int N = Integer.parseInt(NMR[0]);
        int M = Integer.parseInt(NMR[1]);
        int R = Integer.parseInt(NMR[2]);

        int[][] arr = new int[N][M];
        for (int n=0; n<N; n++) {
            StringTokenizer st = new StringTokenizer(sc.readLine());
            for (int m=0; m<M; m++) {
                arr[n][m] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i=0; i<R; i++) {
            for (int j=0; j<Math.min(N, M)/2; j++) {
                int x = j;
                int y = j;
                int temp = arr[x][y];
                int pre = 0;

                for (int a=j+1; a<N-j; a++) {
                    x = a;
                    pre = arr[x][y];
                    arr[x][y] = temp;
                    temp = pre;
                }

                for (int a=j+1; a<M-j; a++) {
                    y = a;
                    pre = arr[x][y];
                    arr[x][y] = temp;
                    temp = pre;
                }

                for (int a=N-j-2; a>j-1; a--) {
                    x = a;
                    pre = arr[x][y];
                    arr[x][y] = temp;
                    temp = pre;
                }

                for (int a=M-j-2; a>j-1; a--) {
                    y = a;
                    pre = arr[x][y];
                    arr[x][y] = temp;
                    temp = pre;
                }
            }
        }

        for (int[] ar : arr) {
            for (int a : ar) {
                System.out.print(a);
                System.out.print(" ");
            }
            System.out.println();
        }
    }
}
